pulsar/producer.go

```diff
@@ -239,7 +239,7 @@ type Producer interface {
 
 	// Flush all the messages buffered in the client and wait until all messages have been successfully
 	// persisted.
-	Flush() error
+	Flush(ctx context.Context) error
```

Comment:
While I agree that we should have taking the context in this method, right now this would be a breaking API change. 

I'd say to add a new `FlushWithContext()` method and, perhaps, mark the other one as deprecated.

Reply:
Just pushed an update. Lmk how it looks!

Btw, do you know if the library will support shutting down via context cancellation  in the future? There's some examples of where we don't do it:

https://github.com/apache/pulsar-client-go/blob/efd9806be9aec8a7ec0e602664bd7dd008822627/pulsar/producer_partition.go#L491

https://github.com/apache/pulsar-client-go/blob/efd9806be9aec8a7ec0e602664bd7dd008822627/pulsar/producer_partition.go#L1470-L1481

As long as things don't take too long to shut down, I suppose it's fine. Ideally, if the context is cancelled, you don't want to wait for a clean/graceful shutdown where we wait for the done channel. We want all the goroutines to terminate right where they are and immediately returned.

Reply:
I would be happy to file an issue if there is not one already.

Reply:
I can't implement `CloseWithCtx` such as in this PR because goroutines might leak (the call to close would terminate before the signal is given on the doneCh, indicating that things have shut down). I think it would be a bigger change.